test/spmd/tensor/test_device_mesh.py

```diff
@@ -0,0 +1,196 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates
+import torch
+
+from torch.distributed.distributed_c10d import (
+    ProcessGroup,
+    GroupMember,
+    new_group,
+)
+from torch.testing._internal.common_utils import run_tests
+from ..test_utils import DistTensorTestBase, with_comms
+from spmd.tensor import DeviceMesh, Tensor, Shard, Replicate
+
+
+class DeviceMeshTest(DistTensorTestBase):
+    @property
+    def world_size(self):
+        return 8
+
+    @with_comms
+    def test_device_mesh_basics(self):
+        # construct a cuda device mesh
+        mesh = DeviceMesh(self.device_type, [0, 1, 2, 3])
+
+        # construct from a cpu local tensor with cuda device mesh
+        # should automatically convert the dist tensor to cuda
+        shard_spec = [Shard(0)]
+        local_tensor = torch.randn(3, 3)
+        dist_tensor = Tensor.from_local(local_tensor, mesh, shard_spec)
+        self.assertEqual(dist_tensor.device.type, self.device_type)
+        self.assertEqual(dist_tensor.local_tensor().device.type, self.device_type)
+
+    @with_comms
+    def test_device_mesh_context_manager(self):
+        with DeviceMesh(self.device_type, list(range(self.world_size))) as mesh:
+            shard_spec = [Shard(0)]
+            local_tensor = torch.randn(3, 3)
+            sharded_tensor = Tensor.from_local(
+                local_tensor, device_mesh=mesh, placements=shard_spec
+            )
+
+        with DeviceMesh(self.device_type, list(range(self.world_size))):
```

Comment:
could we add a test for when a tensor is sharded on the same mesh dimension on 2 tensor dimensions [Shard(0), Shard(0)] and check that the size is correctly computed?

Reply:
In case we have a [Shard(0), Shard(0)] tensor, how are we going to e.g. perform things like all_gather? Would we have to do all_gather on one dimension of the mesh first, then all_gather on the other dimension? Or would it be possible to do a single all_gather on both? In that case I think we'd have a sub process group that covers 2 dimensions of the mesh ?